torch/csrc/deploy/deploy.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+#include <assert.h>
+#include <torch/csrc/deploy/interpreter/interpreter_impl.h>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace torch {
+
+struct MovableObject;
+struct InterpreterManager;
+
+struct TORCH_API InterpreterSession {
+  InterpreterSession(InterpreterSessionImpl* impl, InterpreterManager* manager)
+      : impl_(impl), manager_(manager) {}
+
+  PythonObject self; // when retreived from a PythonMovable this will be set.
+  InterpreterSession(InterpreterSession&&) = default;
+  ~InterpreterSession();
+  PythonObject global(const char* module, const char* name) {
+    return impl_->global(module, name);
+  }
+  PythonObject from_ivalue(at::IValue ivalue) {
+    return impl_->from_ivalue(std::move(ivalue));
+  }
+
+  MovableObject create_movable(PythonObject obj);
+
+ private:
+  friend struct MovableObject;
+  friend struct Package;
+  friend struct InterpreterManager;
+  friend struct MovableObjectImpl;
+  std::unique_ptr<InterpreterSessionImpl> impl_;
+  InterpreterManager* manager_; // if created from one
+  int64_t notify_idx_ = -1;
+};
+
+class TORCH_API Interpreter {
+ private:
+  std::string library_name_;
+  void* handle_;
+  std::unique_ptr<InterpreterImpl> pImpl_;
+
+  InterpreterManager* manager_; // optional if managed by one
+
+ public:
+  Interpreter(InterpreterManager* manager);
+  InterpreterSession acquire_session() const {
+    return InterpreterSession(pImpl_->acquire_session(), manager_);
+  }
+  ~Interpreter();
+  Interpreter(Interpreter&& rhs)
+      : library_name_(std::move(rhs.library_name_)),
+        handle_(rhs.handle_),
+        pImpl_(std::move(rhs.pImpl_)),
+        manager_(rhs.manager_) {
+    rhs.handle_ = nullptr;
+  }
+
+  Interpreter(const Interpreter&) = delete;
+  Interpreter& operator=(const Interpreter&) = delete;
+  Interpreter& operator=(Interpreter&&) = delete;
+  friend struct InterpreterManager;
+};
+
+struct Package;
+
+struct TORCH_API LoadBalancer {
+  LoadBalancer(size_t n) : locks_(new uint64_t[8 * n]), n_(n) {
+    // 8*... to avoid false sharing of atomics on the same cache line
+    memset(locks_, 0, 8 * n_ * sizeof(uint64_t));
+  }
+  void setResourceLimit(size_t n) {
+    n_ = n;
```

Comment:
no bounds check, even in debug mode? we need to have n <= n_ to guarantee that we avoid an out-of-bounds access, right?

Reply:
The only user of this API is checking this, but I added an assert